include/seqan3/io/stream/iterator.hpp

```diff
@@ -166,6 +171,145 @@ class fast_istreambuf_iterator
+     * \{
+     */
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator++()
+    {
+        return *this;
+    }
+    //!\overload
+    fast_ostreambuf_iterator & operator++(int)
+    {
+        return *this;
+    }
+    //!\}
+
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator*()
+    {
+        return *this;
+    }
+
+    //!\brief Writes a character to the associated output stream.
+    fast_ostreambuf_iterator & operator=(char_t const c)
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+        {
+            if (stream_buf->sputc(c) == traits_t::eof()) // overflow() [virtual], then write character
+                throw std::ios_base::failure{"Cannot write to output stream (reached traits::eof() condition)."};
+        }
+        else
+        {
+            *stream_buf->pptr() = c;
+            stream_buf->pbump(1); // advance pptr() in put area without any checks
+        }
+        return *this;
+    }
+
+    //!\brief Returns `true if this iterator has encountered the end-of-file condition on output, `false` otherwise.
+    bool failed() const noexcept
+    {
+        return stream_buf->overflow() == traits_t::eof();
+    }
+
+    /*!\brief Writes a range to the associated output.
+     * \tparam range_type The type of range to write; Must model std::ranges::forward_range.
+     * \param[in] rng The range to write.
+     */
+    template <std::ranges::forward_range range_type>
+    auto write_range(range_type && rng)
+    {
+        size_t buffer_space = stream_buf->epptr() - stream_buf->pptr();
+
+        auto rng_it = std::ranges::begin(rng);
+        auto rng_end = std::ranges::end(rng);
+
+        if constexpr (std::ranges::sized_range<range_type>)
+        {
+            auto characters_to_write = std::min<size_t>(std::ranges::size(rng), buffer_space);
+            auto copy_res = std::ranges::copy_n(rng_it, characters_to_write, stream_buf->pptr());
+            rng_it = copy_res.in;
+            stream_buf->pbump(characters_to_write);
+        }
+        else
+        {
+            for (size_t i = 0; rng_it != rng_end && i < buffer_space; ++rng_it, ++i)
+            {
+                *stream_buf->pptr() = *rng_it;
+                stream_buf->pbump(1);
+            }
+        }
+
+        if (rng_it == rng_end) // no more characters to write
+        {
+            if constexpr (std::ranges::viewable_range<range_type>)
+                return rng_it;
+            else
+                return;
+        }
+
+        // Push one more character and flush
+        if (stream_buf->overflow(*rng_it) == traits_t::eof())
+            throw std::ios_base::failure{"Cannot write to output stream (reached traits::eof() condition)."};
+
+        ++rng_it; // drop 1 character that has been written in overflow()
+
+        // write remaining range recursively
+        return write_range(std::ranges::subrange<decltype(rng_it), decltype(rng_end)>{rng_it, rng_end});
```

Comment:
```suggestion
        if constexpr (std::ranges::sized_range<range_type> &&
                      !std::sized_sentinel_for<std::ranges::sentinel_t<range_type>, std::ranges::iterator_t<range_type>>)
            return write_range(std::ranges::subrange<decltype(rng_it), decltype(rng_end), size_t>{rng_it, rng_end, size_left - 1});
        else
            return write_range(std::ranges::subrange<decltype(rng_it), decltype(rng_end)>{rng_it, rng_end});
```

Reply:
How can a range be sized but the sentinal type not model `sized_sentinal_for` for the iterator?

Reply:
Oh for view take.. :sleepy: